src/linker.rs

```diff
@@ -229,354 +266,169 @@ pub struct LinkerOptions {
 /// BPF Linker
 pub struct Linker {
     options: LinkerOptions,
-    context: LLVMContextRef,
-    module: LLVMModuleRef,
-    target_machine: LLVMTargetMachineRef,
+    context: LLVMContext,
     diagnostic_handler: DiagnosticHandler,
 }
 
 impl Linker {
     /// Create a new linker instance with the given options.
-    pub fn new(options: LinkerOptions) -> Self {
-        Linker {
+    pub fn new(options: LinkerOptions) -> Result<Self, LinkerError> {
+        let (context, diagnostic_handler) = llvm_init(&options);
+
+        Ok(Linker {
             options,
-            context: ptr::null_mut(),
-            module: ptr::null_mut(),
-            target_machine: ptr::null_mut(),
-            diagnostic_handler: DiagnosticHandler::new(),
-        }
+            context,
+            diagnostic_handler,
+        })
+    }
+
+    /// Link and generate the output code to file.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// # use std::{collections::HashSet, path::Path};
+    /// # use bpf_linker::{Cpu, Linker, LinkerInput, LinkerOptions, OptLevel, OutputType};
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// let path = Path::new("/path/to/object-or-bitcode");
+    /// let bytes: &[u8] = &[]; // An in memory object/bitcode
+    /// # let options = LinkerOptions {
+    /// #     target: None,
+    /// #     cpu: Cpu::Generic,
+    /// #     cpu_features: String::new(),
+    /// #     optimize: OptLevel::Default,
+    /// #     unroll_loops: false,
+    /// #     ignore_inline_never: false,
+    /// #     llvm_args: vec![],
+    /// #     disable_expand_memcpy_in_order: false,
+    /// #     disable_memory_builtins: false,
+    /// #     allow_bpf_trap: false,
+    /// #     btf: false,
+    /// # };
+    /// # let linker = Linker::new(options)?;
+    ///
+    /// // Link to a file
+    /// linker.link_to_file(
+    ///     vec![
```

Comment:
this is why Vec is bad: you're forcing people to allocate here, while

```
[
	LinkerInput::try_from(path),
]
```

should work too

something like ```impl IntoIterator<Item = impl BorrowMut<LinkerInput<'a>>>```
should work